### PR TITLE
[llvm-exegesis] Kill process that recieve a signal

### PR DESCRIPTION
**Pull request by boomanaiden154**
Before this patch, llvm-exegesis would leave processes lingering that experienced signals like segmentation faults. They would up in a signal-delivery-stop state under the ptrace and never exit. This does not cause problems (or at least many) in llvm-exegesis as they are cleaned up after the main process exits, which usually happens quickly. However, in downstream use, when many blocks are being executed (many of which run into signals) within a single process, these processes stay around and can easily exhaust the process limit on some systems.

This patch cleans them up by sending SIGKILL after information about the signal that was sent has been gathered.

### CODE REVIEW

**Comment by boomanaiden154** [comment]
Fixes https://github.com/google/gematria/issues/76.

File: llvm/tools/llvm-exegesis/lib/BenchmarkRunner.cpp [hunk 1]
```diff
@@ -342,7 +342,7 @@ class SubProcessFunctionExecutorImpl
       return make_error<Failure>("Failed to attach to the child process: " +
                                  Twine(strerror(errno)));
 
-    if (wait(NULL) == -1) {
+    if (waitpid(ParentOrChildPID, NULL, 0) == -1) {
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/lib/BenchmarkRunner.cpp:345]: At that point `ParentOrChildPID` is always the parent, but it's not immediately obvious (I had to scroll up). What about restructuring this function to have:

```

[[noreturn]] void runChildSubprocess(int ReadFD, int FriteFD) {
   // We are in the child process, close the write end of the pipe.
      close(PipeFiles[1]);
      // Unregister handlers, signal handling is now handled through ptrace in
      // the host process.
      sys::unregisterHandlers();
      prepareAndRunBenchmark(PipeFiles[0], Key);
      llvm_unreachable("Child process didn't exit when expected.");
}

Error runParentSubprocess(pid_t PID, int ReadFD, int FriteFD) {
    const ExegesisTarget &ET = State.getExegesisTarget();
    ...
}


createSubProcessAndRunBenchmark() {
  ...
  if (ParentOrChildPID == -1) {
    ...
  }
  
  if (ParentOrChildPID == 0) {
     runChildSubprocess(PipeFiles[0], Pipefiles[1]);
      llvm_unreachable("Child process didn't exit when expected.");
  }

  return runParentSubprocess(ParentOrChildPID, PipeFiles[0], Pipefiles[1]);
}
```

**Review by legrosbuffle — APPROVED** [review]

File: llvm/tools/llvm-exegesis/lib/BenchmarkRunner.cpp (hunk 1 above)
Comment (boomanaiden154) [reply llvm/tools/llvm-exegesis/lib/BenchmarkRunner.cpp:345]: That's a good point. The suggested refactoring would definitely make the code cleaner. I'm going to land this PR and then open up another PR with the suggested refactoring to try and keep the history cleanish. Thanks for the suggestion!

Comment (boomanaiden154) [reply llvm/tools/llvm-exegesis/lib/BenchmarkRunner.cpp:345]: See https://github.com/llvm/llvm-project/pull/86232.

**Pull request merged**